오픈마켓 [Step 1] 산체스

sanches37 opened the pull request:
@GREENOVER 
안녕하세요 그린! 오픈마켓 step1 PR을 보냅니다. 
지난 오픈마켓 프로젝트에서 step1에 시간을 많이 사용했었습니다.
현재코드에 부족한 부분이 많지만, 다음 단계에 시간을 투자해보고 싶어서 지난 프로젝트 step1과 코드가 많이 바뀌지 않았습니다.

고민한점
1. Request 타입의 createRequest에 같은 protocol을 채택한 타입을 바꿔 끼워가며 사용하는 구조입니다.
메서드 내에 if문으로 분기하는 부분으로 인해 코드가 길어지고 있어서, 코드를 줄이기 위해 protocol 구조를 바꿔보려했지만, 
깔끔한 구현이 되지 않아서 기존의 코드에서 수정하지 못했습니다.

2. NetworkManager 가 너무 많은 것을 하고 있다고 생각하여 Request 관련된 코드를 타입을 분리하여 구현했습니다.

3. 테스트를 위한 Mock 관련 코드가 기존에는 OpenMarket 타겟으로 있었는데, OpenMarketTests 타겟이 되도록 구현했습니다

부족한 부분이 많을 것으로 생각되는데 많은 리뷰 잘 부탁드립니다.

sanches37 committed:
feat: 데이터를 활용할 모델타입 구현
struct ItemsData 구현
struct ItemData 구현
Snake-case 로 변환할 codingKey Enum 구현

sanches37 committed:
feat: 디코딩에 사용할 ParsingManager 구현
ParsingManager 내 receivedDataAsset 구현
ParsingManager 내 decodedJsondata 구현

sanches37 committed:
feat: Mock 데이터 저장
프로젝트 문서에서 Items.json 파일과 Item.json 파일을 xcode 에 저장

sanches37 committed:
refactor: PasingManager 내 매서드 static 삭제
인스턴스 후 사용하기 위해 static 삭제

sanches37 committed:
test: ItemsData 테스트코드 구현
ItemsData 성공 테스트 구현
ItemData 타입 내 타입 및 스펠링 오류 수정

sanches37 committed:
test: ItemData 타입을 사용한 디코딩 테스트 구현

sanches37 committed:
feat: Network Request에 사용할 Enum 타입들 구현
APIMethod enum 구현
APIURL enum 구현
ContentType enum 구

sanches37 committed:
feat: Network Request에 사용할 RequestProtocol 구현

sanches37 committed:
feat: Network 에 사용할 Request 구현
Request struct 내 boundary 프로퍼티 구현
Request struct 내 createBody 메서드 구현
image 를 전달할 ImageData 구현

sanches37 committed:
refactor: ImageData 내 누락 된 코드 추가
self.mimeType = mimeType 코드 추가

sanches37 committed:
refactor: Result 타입 내 createBody 수정
body.append("--\(boundary)\(lineBreak)") 로 표기 오류로 수정

sanches37 committed:
refactor: Requestable 프로토콜 수정
httpBody 를 가진 API에 채택하기 위해 RequestableWithHttpBody 추가

sanches37 committed:
feat: RequestAPI struct 추가
Request 사용시 API에 따라 갈아낄 수 있는 RequestAPI 추가
requestProtocol 내 RequestableWithMultipartForm 추가
requestProtocol so RequestableWithHttpBody 수정

sanches37 committed:
feat: 전체 API 의 Request 를 전달하는 createRequest 구현
Request 타입 내 createRequest 구현
RequestError 구현
RequestProtocol 수정
ContentType enum 내 httpHeaderField static 프로퍼티 추가

sanches37 committed:
feat: ParameterData 모델 구현
Post, Patch, Delete 에 사용할 ParameterData 모델 구현
MultipartFormData struct 구현
DeleteParameterData struct 구현

sanches37 committed:
feat: NetworkManager 타입 구현
NetworkManager 내 commuteWithAPI 구현
NetworkError 구현
RequestError 를 NetworkError 로 통합

sanches37 committed:
feat: 네트워크 무관테스트를 진행할 Mock 객체 구현
MockURLSession 타입 구현
MockURLSessionDataTask 타입 구현
NetworkProtocol.swift 내 URLSessionProtocol 구현

sanches37 committed:
refactor: Mock 객체 적용할 수 있게 NetworkManager 수정
테스트를 위해 init 에 session, valuableMethod 를 다르게 적용할 수 있게 구현
에러메시지를 전달할 수 있도록 코드 수정

sanches37 committed:
test: 네트워크 무관테스트 적용

sanches37 committed:
test: get, post, patch, delete 를 확인하는 코드추가
viewDidLoad 에 코드 추가

sanches37 committed:
refactor: RequestAPI 변수를 상수로 변경
정렬 및 띄어쓰기 수정

sanches37 committed:
feat: SwiftLint 적용
SwiftLint 적용하기 위해 코드 수정
imageData struct -> Media 로 이름 변경

GREENOVER commented on OpenMarket/.swiftlint.yml:
```diff
@@ -0,0 +1,10 @@
+excluded:
+- Pods
+- OpenMarket/AppDelegate.swift
+- OpenMarket/SceneDelegate.swift
```
크 SwiftLint를 적용해보셨네요👍
excluded 처리까지 꼼꼼히 잘해주셨네요 최고입니다!
실제로 SwiftLint를 통해 통일된 코드 컨벤션이 있으면 유지보수하거나 새로운 개발자가 와도 코드 이해가 쉽겠죠🥳

sanches37 replied:
메모장 프로젝트 요구사항에 있어서, 야곰께서 오픈마켓에도 적용해보는게 어떻게냐고 하셨습니다
SwiftLint를 적용해보면서 코드 컨벤션에 관해서 알게 되었네요. 😄

GREENOVER replied:
Apple에서 제공하는 https://github.com/apple/swift-format 도 있으니 한번 보시면 좋을것 같아요👍

sanches37 replied:
공유해주신 자료 감사합니다🙏🏼

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -300,6 +339,70 @@
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "${PODS_ROOT}/SwiftLint/swiftlint\n";
```
빌드 파서에 쉘 스크립트도 잘 적용하셨네요👍
만약 코드가 방대해진다면 SwiftLint Rule중에 Supports autocorrection 라는 좋은것도 있습니다~!
스크립트에 같이 붙여준다면 편리하게 사용할 수 있을거에요😁

sanches37 replied:
감사합니다 Supports autocorrection 대해서도 공부해보고 적용할 수 있도록 해보겠습니다. ☺️

GREENOVER commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -11,33 +11,28 @@ class ViewController: UIViewController {
         
-        
     // getItem
 //        networkManager.commuteWithAPI(API: GetItemAPI(id: 643)) { _ in
```
실제 네트워킹 테스트를 하다 깜빡하고 주석처리만 되고 지워지지 않은 코드들이 보이네요😭
완성된 PR에는 사용되지 않는 코드나 불필요한 주석은 제거하는편이 좋습니다🔥

sanches37 replied:
그린께서 코드 리뷰를 하실 때 주석을 살려서 테스트해보시면 좋을 것 같아서 남겨두었었는데, 리뷰 수정을 하면서 지우도록 하겠습니다.

GREENOVER replied:
아하 깊은뜻이..!! 감동이네요😀
테스트에 대해 알리고 싶다면 PR 내용에 테스트 방법을 기재하는것도 좋은 방향이 될것 같아요ㅎㅎ
저보다 산체스가 한발 앞서갔네요!
그래도 코드는 깔끔한게 좋겠죠🧐

GREENOVER commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -11,33 +11,28 @@ class ViewController: UIViewController {
-        
+
     // getItems
         networkManager.commuteWithAPI(API: GetItemsAPI(page: 1)) { _ in
```
파라미터명이 API로 모두 대문자인데 API,ID,URL같이 규정된 명칭에 대해서 대문자로 잘 통일해주셔서 사용한게 보입니다!👍
그런데 RequestAPI 파일 내 url 프로퍼티는 소문자로 되어있는데 조금 더 통일되면 좋을것같아요👍
저같은 경우라면 변수라는 가정에선 소문자로 나타내고 합성어가 된다면 예를들어, dataOfURL같이 표현하면서 가급적 스위프트의 카멜케이스를 맞춰주려 노력해요🏃🏻
네이밍이 정답이 있는건 아니라 산체스가 변수명일때 대문자로 통일한다고 규정했으면 다른쪽도 통일해주면 좋을것 같습니다!
더 좋은 의견있으시면 알려주세요📣

sanches37 replied:
맞습니다. 변수명 통일이 잘 되어있지 않았네요. 
그린이 선호하시는 방법대로 규정된 명칭도 소문자로 나타내고, 합성어가되었을 때만 대문자를 사용하는 방법이 저도 좋아보이네요.
네이밍을 수정하도록 하겠습니다. 😃

sanches37 replied:
규정된 명칭이 앞쪽에 오는 경우 ApiMetod 가 아니라 APIMethod 맞는거겠죠? URLSession 을 참고해서 APIMethod 로 하겠습니다.😊

GREENOVER commented on OpenMarket/OpenMarket/Network/Request.swift:
```diff
@@ -35,7 +35,7 @@ struct Request {
+    private func createBody(params: [String: Any]?, image: [Media]?) -> Data {
         var body = Data()
         
         let lineBreak = "\r\n"
```
반복되는 문자열을 따로 상수 처리하여 사용한점 💯💯💯
지금도 충~분히 훌륭하지만 조금 더 욕심을 내보자면 "\r\n\r\n"일때는 문자열을 +로 합쳐주었는데
만약 "\r\n\r\n\r\n\r\n과 같다면 코드 가독성이 조금 떨어질 수도 있을것 같아요!
한번만 사용된다면 괜찮지만 "\r\n\r\n"도 상수 처리해주면 어떨까요?
반복되는 부분은 최대한 세세하게 상수처리하는것에 대해 산체스는 어떤 의견이 있는지 궁금하네요🧐

sanches37 replied:
let doubleLineBreak = "\r\n\r\n" 를 만들어서 두줄 띄어쓰기를 하는 상황에 사용하도록 수정했습니다.
lineBreak + lineBreak 를 doubleLineBreak 로 사용하게 되어서 더 명시적이고 보기 쉬워졌다고 생각합니다. 감사합니다 👍🏼

GREENOVER commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -11,33 +11,28 @@ class ViewController: UIViewController {
-        
+
     // getItems
         networkManager.commuteWithAPI(API: GetItemsAPI(page: 1)) { _ in
```
네트워크 통신을 따로 메서드로 분리해서 호출해주면 어떨까요?
메서드로 분리하다보면 나중에 추가 구현사항이 있거나 할때 구분이 잘되서 용이할거에요🙌

GREENOVER replied:
해당 부분에서는 실제 데이터를 가져와서 아무 처리를 안해주고 있습니다!
사용하신다면 클로저를 구현해줘야할것 같아요👏

sanches37 replied:
현재 테스트용으로 클로저를 사용한 부분이라, 다음 단계에서 네트워크 호출을 사용할 때 메서드로 분리해서 호출해 보겠습니다! 😀

GREENOVER commented on OpenMarket/OpenMarket/Model/imageData.swift:
```diff
@@ -0,0 +1,42 @@
+
+import UIKit
+
+enum MimeType: CustomStringConvertible {
```
CustomStringConvertible 프로토콜을 채택하면 어떤 이점이 있나요❓❓

sanches37 replied:
CustomStringConvertible를 채택하면 호출할 때  MImeType.프로퍼티명으로 접근하지 않고 MImeType 을 쓰는것으로 description 프로퍼티에 접근해서 사용할 수 있어서 편리한 이점이 있습니다 😃

GREENOVER commented on OpenMarket/OpenMarket/Model/imageData.swift:
```diff
@@ -0,0 +1,42 @@
+
+struct imageData {
+    let key: String = "images[]"
+    let filename: String
```
filename이 저는 파일+이름이 합쳐진 합성어 같은 느낌이 드는데 카멜케이스를 적용해볼수 있을것 같아요🐫

sanches37 replied:
멀티파트폼으로 전달할시에 filename 으로 전달을 해서, 상수 이름에서도 카멜케이스를 고려하지 않고 사용한 것 같습니다. 
카멜케이스를 고려하여 수정하겠습니다😀

GREENOVER commented on OpenMarket/OpenMarket/Model/ItemData.swift:
```diff
@@ -0,0 +1,27 @@
+
+import Foundation
+
+struct ItemData: Codable {
```
해당 구조체 자체가 Item의 Data 즉 정보들을 담고 있는데 조금 더 추상적인 네이밍은 어떨까요?

sanches37 replied:
정보들을 담고 있는데 Data를 붙여줄 필요가 없을 것 같아서 수정했습니다. 네이밍도 야아마켓에서 판매하는 것에 관한 정보니까 Product 로 수정해봤습니다. ImageData 와 ParameterData 또한 Data를 제외하고 메서드명 파일명을 수정했습니다. 😀

GREENOVER commented on OpenMarket/OpenMarket/Model/ItemsData.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+struct ItemsData: Codable {
```
요기두 ItemsData만으로는 ItemData와 같이 가독성이 떨어지는 부분이 있는데 더 좋은 네이밍이 있을것 같아요🧐

sanches37 replied:
Products 로 하면 Product 와 네이밍이 비슷해서 ProductCollection 으로 네이밍을 변경하였습니다.

GREENOVER replied:
제가 의도한 바를 정확히 파악하셨군요 최고👍

GREENOVER commented on OpenMarket/OpenMarket/Assets.xcassets/test1.imageset/Contents.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "test1.png",
```
이미지도 테스트를 해보시려고 넣으셨군요☺️
테스트 파일이여도 명확한 이름이면 좋습니다! 
저희는 누군가가 아니더라도 미래의 자신과 협업을 하니까요🙏🏻

sanches37 replied:
나중에 제가 봤을 때 어떤 의미로 어디서 사용했는지 알 수 있게 Step1PostTestImage 로 이름을 변경해봤습니다. 감사합니다.😆

GREENOVER commented on OpenMarket/OpenMarket/Extension/BaseTypeExtension.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+extension Data {
```
👏👏👏커스텀이 필요한것을 익스텐션 파일로 빼서 다뤄주신 부분 최고👍🏻
조금 욕심을 내보자면 BaseTypeExtension만 보고 어떤 느낌인지 감이 안와요😭
Data+extension과 같이 조금 더 명확하면 어떨까요?

sanches37 replied:
말씀해주신 방법으로 수정하면 다른 타입의 extension을 추가해도 Int+extension , String+extension 처럼 적용해서 훨씬 눈에 들어오고 찾기 좋을 것 같습니다. 수정하겠습니다 😆

GREENOVER commented on OpenMarket/OpenMarket/Extension/BaseTypeExtension.swift:
```diff
@@ -0,0 +1,16 @@
+import Foundation
+
+extension Data {
+    mutating func append(_ string: String) {
```
만약 append 메서드를 호출할때 아래와 같이 익스텐션되서 새로 정의된 기능을 하고싶지 않을땐 어떡하죠🥲

sanches37 replied:
<img width="427" alt="스크린샷 2021-09-02 오후 12 45 27" src="https://user-images.githubusercontent.com/84059338/131778634-b17b142b-5028-45b6-a353-7baaef70cd0c.png">

append 메서드를 호출할 때 같은 메서드명의 다른 매개변수를 선택할 수 있어서 괜찮다고 생각했습니다

GREENOVER replied:
크 완벽히 이해하고 계시군요!
이부분은 더 말씀드릴 부분이 없네요 최고👍👍👍

GREENOVER commented on OpenMarket/OpenMarket/Manager/NetworkManager.swift:
```diff
@@ -0,0 +1,77 @@
+class NetworkManager {
+    private let request = Request()
+    private let session: URLSessionProtocol
+    private var valuableMethod: [APIMethod] = []
```
valuableMethod가 어떤의미인가요❓

sanches37 replied:
valuableMethod는 HTTPMethod 가 제대로 전달됐는지 테스트하기 위한 용도로 만들었는데 Method만 적어서 의미가 부정확한 것 같습니다. valuableHTTPMethod 로 네이밍을 변경했습니다. 😊

GREENOVER replied:
훨씬 좋아요! 근데 valuable이 가치있는 이란 뜻으로 저는 해석이 되어서 좀 모호한 느낌으로 말씀드렸었어요!
다른 뜻이 있을까요? (제가 영포자라...😭)

sanches37 replied:
제가 잘못 사용한 것 같습니다 가치있는HTTPMethod 로 네이밍을 정한거군요🤣. 의미에 맞게 네이밍 수정하겠습니다.

sanches37 replied:
적절할지 모르겠지만 적용가능하다는 의미로 applicableHTTPMethod 로 수정했습니다 😊

GREENOVER commented on OpenMarket/OpenMarket/Manager/NetworkManager.swift:
```diff
@@ -0,0 +1,77 @@
+    private static let rangeOfSuccessState = 200...299
+    
+    init(session: URLSessionProtocol = URLSession.shared,
+         valuableMethod: [APIMethod] = APIMethod.allCases) {
```
valuableMethod의 초기값을 지정해주었다면 위 StoredProperty로 지정되는것하고
```swift
private var valuableMethod: [APIMethod]
```
로 타입 선언만 해주는것하고 어떤 차이가 있을까요❓❓

sanches37 replied:
기본값으로 APIMethod.allCases 을 가지고 있기 때문에 빈배열을 만들지 않고 타입만 정의해줘도 동일하게 작동하게 되는 것 같습니다. 감사합니다😆

GREENOVER commented on OpenMarket/OpenMarket/Manager/NetworkManager.swift:
```diff
@@ -0,0 +1,77 @@
+            return
+        }
+        session.dataTask(with: request) { data, response, error in
+            guard error == nil else {
```
error 처리에 대해서 if let도 있을텐데 guard let을 사용하여 처리를 한 이유가 궁금해요🧐

sanches37 replied:
처음에 if let error = error 로 처리하고 failure(error) 로 처리하는 구조로 진행했었습니다.
에러메시지를 확실히 하기 위해 NetworkError로 처리하려고 구조를 바꾸니
if let _ = error 구조가 되어서 언더스코어를 사용하지 않기 위해 guard error = nil  일때로 가드문을 사용했습니다.
if let 구문이 더 깔끔해서 사용하려했는데 언더스코어가 들어가더라도 if let _ = error 로 고쳐보는게 좋을까요?

GREENOVER replied:
저는 코드 문맥상 if let이 더 맞는 표현같습니다🧐

sanches37 replied:
감사합니다. if let을 사용해서 수정해보겠습니다

GREENOVER commented on OpenMarket/OpenMarket/Manager/NetworkManager.swift:
```diff
@@ -0,0 +1,77 @@
+        self.valuableMethod = valuableMethod
+    }
+    
+    func commuteWithAPI(API: Requestable, completionHandler: @escaping(Result<Data, Error>) -> Void) {
```
멋지게 만들어주신 에러타입을 가지고 핸들러의 탈출 클로저 부분을 typealias를 통해 선언해두면 아래에서 조금 더 코드를 줄일 수 있을것 같습니다👍🏻
```swift
completionHandler(.failure(.requestFailed)
```
처럼만 호출 할 수 있을거에요🧐
물론 지금도 충분히 에러 처리가 잘 되어있습니다👏

sanches37 replied:
말씀해주신대로 메서드내에서 NetworkError 타입만을 실패처리로 사용하고 있어서 failure 타입을 NetworkError 로 변경했습니다. 😆

GREENOVER commented on OpenMarket/OpenMarket/Manager/NetworkManager.swift:
```diff
@@ -0,0 +1,77 @@
+                completionHandler(.failure(NetworkError.responseFailed))
+                return
+            }
+            debugPrint(response)
```
debugPrint로 로그 파악을 하는것 좋네요👍🏻
로그 추적을 위해 print와 debugPrint는 어떤 차이가 있을까요?
로그 파악 테스트 시에 해당 코드를 심어주었던것이면 커밋할때는 빼주는게 좋을것 같아요☺️

sanches37 replied:
debugPrint가 읽는데는 불편할 순 있지만 상세내용까지 출력되는 것으로 알고 있습니다. 
찾아보니 debug 용도로는 애플에서도 debugPrint를 추천하고 있다고 하네요 😄

GREENOVER commented on OpenMarket/OpenMarket/Manager/ParsingManager.swift:
```diff
@@ -0,0 +1,43 @@
+import UIKit
+
+enum ParsingError: Error, LocalizedError {
+    case assetFailed
```
직역하면 에셋실패인데 에셋이 뭐가 실패인지 조금 불명확한것 같아요🥲
에셋을 불러오는 실패라는 명확한 케이스 네이밍이 있으면 좋겠습니다🙌

sanches37 replied:
에셋을 로드하는데 실패했다는 네이밍이 어울리겠네요 assetLoadFailed 로 네이밍을 변경했습니다. 😌

GREENOVER commented on OpenMarket/OpenMarket/Manager/ParsingManager.swift:
```diff
@@ -0,0 +1,43 @@
+struct ParsingManager {
+    private let jsonDecoder = JSONDecoder()
+    
+    func receivedDataAsset(assetName: String) throws -> NSDataAsset {
```
함수명이 명사의 느낌이 들어서 조금 어색해요🥺
어떤 기능을 하는지를 포함한 더 좋은 네이밍이 있을거라 생각합니다 :D

sanches37 replied:
에셋의 데이터를 불러오는 의미로 LoadedDataAsset 으로 네이밍을 변경했는데 괜찮을까요? 🙂

GREENOVER replied:
훨씬 가독성이 좋습니다👍
그런데 위에 assetLoadFailed도 LoadAssetFailed와 같은 메서드명으로 통일은 어떨까요?
동작을 나타내는 Load라는게 먼저 나오는게 더 좋을것 같아요

sanches37 replied:
loadAssetFailed 로 수정했습니다. LoadedDataAsset 도 실수로 카멜케이스를 적용하지 않아서 loadedDataAsset 로 다시 수정했습니다 😅

GREENOVER commented on OpenMarket/OpenMarket/Model/ItemData.swift:
```diff
@@ -0,0 +1,27 @@
+    let descriptions: String?
+    let price: Int
+    let currency: String
+    let stock: Int
```
수량같은 경우 음수의 값을 가질 경우가 있을까요🧐
선택이긴 하지만 price와 stock에서 Int와 UInt 어떤것을 택할지도 상황에 따라 고민해볼 수 있을것 같아요ㅎㅎ
(물론 UInt가 맞아! 라는 뜻은 절대절대 아닙니다!)

sanches37 replied:
<img width="645" alt="스크린샷 2021-09-02 오후 1 03 05" src="https://user-images.githubusercontent.com/84059338/131780062-87769cdc-6bf4-46f6-bcb3-36993c60bab8.png">

음수를 사용하지 않을 때는 UInt를 고려했었는데 공식문서의 내용을 보고, 음수가 아닐때도 Int를 사용하게 됐습니다.

GREENOVER replied:
공식문서를 통해 의견에 대한 근거를 제시하시는 모습 멋집니다👍👍

GREENOVER commented on OpenMarket/OpenMarket/Network/NetworkProtocol.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+protocol URLSessionProtocol {
```
프로토콜로 추상화 시키는거 너무 너무 세련스럽습니다😀
저보다 나은데요?💯💯💯

sanches37 replied:
우아한 형제들 기술블로그를 참고해서 그렇습니다 😂

GREENOVER commented on OpenMarket/OpenMarket/Network/RequestAPI.swift:
```diff
@@ -0,0 +1,67 @@
+
+import UIKit
+
+struct GetItemAPI: Requestable {
```
크.. Requestable 프로토콜을 만들어 놓고 이렇게 활용하시는것까지 최곱니다👍🏻👍🏻👍🏻

GREENOVER replied:
아아! Struct에 대해 잘 이해하고 활용하시는것도 정말 베스트입니다ㅎㅎ

sanches37 replied:
감사합니다 🙏🏼

GREENOVER commented on OpenMarket/OpenMarket/Network/RequestEnum.swift:
```diff
@@ -0,0 +1,63 @@
+
+import Foundation
+
+enum APIMethod: CustomStringConvertible, CaseIterable {
```
CustomStringConvertible 프로토콜을 사용해준 이유가 궁금합니다🧐

sanches37 replied:
string 타입에 적용하다보니 APIMethod.description 으로 적용하고 있습니다. CustomStringConvertible 사용하지 않고 프로퍼티 이름을 명시해주고 사용하는게 좋을 것 같아서 코드를 수정했습니다. ContentType 도 description으로 불러오고 있어서 수정했습니다
APIHTTPMethod 로 네이밍도 변경했습니다 🙂

GREENOVER commented on OpenMarket/OpenMarket/Network/RequestProtocol.swift:
```diff
@@ -0,0 +1,19 @@
+
+import Foundation
+
+protocol Requestable {
```
프로토콜 사용 최고최고🎉🎉🎉

sanches37 replied:
😄😆

GREENOVER commented on OpenMarket/OpenMarketTests/Mock/Item.json:
```diff
@@ -0,0 +1,17 @@
+{
```
에셋에 Mock 데이터 파일을 담는거와 OpenMarketTests > Mock 폴더에 담는건 어떤 차이가 있을까요🙋🏻‍♂️

sanches37 replied:
OpenMarketTests Mock 폴더에 담는 건 타겟을 OpenMarketTests 으로 하는 차이가 있을 것 같습니다. 🙂

GREENOVER commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,101 @@
+        parsingManager = ParsingManager()
+    }
+    
+    func test_Items_에셋파일을_ItemsData_타입으로_디코딩에_성공하면_title은_MacBookPro다() {
```
오호 매번 영어로된 메서드명만 보다 한국에 맞는 로컬라이즈된 메서드명을 보는것 같아요!😁
그런데 팀원이 한국어를 잘 못하면 어떡하죠?ㅠㅠ
자세한 설명같은 메서드명으로 저는 이해가 정말 잘되었지만 어떤 기능을 하는지에 중점을 둔 메서드명이 어떨까요?
"디코딩에 성공하면 타이틀은 맥북프로다" 라는 설명은 코드를 보고 이해하도록 하면 좋을것 같아요☺️

sanches37 replied:
메서드명을 한글로 하면 그런 단점이 있는 것 같습니다. 메서드를 영어로 하더라도 코드를 보고 이해할 수 있도록 테스트를 짤 수 있게 노력하겠습니다. 추후 테스트를 작성할 땐 말씀하신 부분 고려해서 구현하겠습니다 👍🏼

GREENOVER commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -0,0 +1,101 @@
+        // when
+        guard let jsonData = try? parsingManager?.receivedDataAsset(assetName: expectInputValue),
+              let decodedData = try? parsingManager?.decodedJsonData(type: ItemData.self, data: jsonData.data) else {
+            return XCTFail()
```
요부분 테스트케이스에 통과하지 못하는데 의도인가요🧐

sanches37 replied:
expect 값을 바꿔가며 테스트를 시험해보다가 expectInputValue 값을 잘못 넣었습니다.
let expectInputValue = "Item" 으로 변경하여 테스트에 통과하도록 수정했습니다. 꼼꼼하게 확인하지 못했었네요.🥲

GREENOVER commented on OpenMarket/OpenMarket/Network/RequestEnum.swift:
```diff
@@ -0,0 +1,63 @@
+    var description: String {
+        switch self {
+        case .getItems:
+            return Self.baseURL + "items/"
```
Enum의 associated values를 잘 이용하면 네트워크 리슨받는곳에서 id를 넘겨주지 않고 해당 switch 분기에서 처리하는 방법도 있습니다☺️
추후 진행사항을 볼때 어느 방향이 더 좋을까요?

sanches37 replied:
이 부분은 고민은 한 부분인데, 다음 스탭에서 이부분을 사용할 때 Associated Values 를 사용하는 것으로 구현해보겠습니다 💪🏻

GREENOVER commented on OpenMarket/OpenMarket/Manager/ParsingManager.swift:
```diff
@@ -0,0 +1,43 @@
+        return dataAsset
+    }
+    
+    func decodedJsonData<T: Decodable>(type: T.Type, data: Data) throws -> T {
```
Json도 네이밍 통일성을 위해 JSON으로 맞춰주는것도 권유드립니다🙋🏻‍♂️

sanches37 replied:
decodedJSONData로 변경했습니다. 감사합니다😀

GREENOVER commented on OpenMarket/OpenMarket/Model/ImageData.swift:
```diff
@@ -0,0 +1,44 @@
+        case .jpeg:
+            guard let data = image.jpegData(compressionQuality: 0.7) else { return nil }
+            self.data = data
+            self.filename = "\(arc4random()).jpeg"
```
fileName이 난수로 지정되어있는데 운이 나빠서 동일한 난수가 추출되면 파일명은 중복이 되는건가요?
정말 유니크한 파일명을 추출하고 싶다면 Date()를 사용해도 좋아요!

sanches37 replied:
이런 방법이 있었군요! 🙌 너무 좋은 방법을 배운 것 같습니다.  파일을 올린 날짜도 파일명만으로 알수도 있구요. 감사합니다 👍🏼

GREENOVER commented on OpenMarket/OpenMarketTests/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,34 @@
+    
+    func dataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
+        let successResponse = HTTPURLResponse(url: url, statusCode: 200, httpVersion: "2", headerFields: nil)
+        let failureResponse = HTTPURLResponse(url: url, statusCode: 402, httpVersion: "2", headerFields: nil)
```
402 에러 상태코드로 지정한 이유가 있을까요?🧐

sanches37 replied:
402 에러로 지정한 이유는 따로 없고, 테스트를 위해 200...299 상태코드에서 벗어나는 코드를 사용했습니다. 😀

GREENOVER reviewed (commented):
안녕하세요 산체스 너무너무 반갑습니다!🙋🏻‍♂️
앞으로 3주간 긴 여정을 저도 리뷰어로써 잘 부탁드리며 함께 완수해봐요☺️
혼자 오픈마켓을 하게되셨는데 고군분투하시며 많은 부분을 고민하셨던 흔적이 보여서 대단하고 멋집니다!
특히 코드를 읽으며 이해하는데 너무너무 편했습니다👏
제가 네이밍 집착증도 많고 산체스와 함께 프로젝트를 하는 느낌을 받다보니 코멘트가 다소.. 아니 많습니다😭
훌륭한 산체스의 코드에 "이런 방향은 어떨까요?" 하는 의견들이니 모두 다 받아들이시지 않으셔도 됩니다!
산체스의 의도가 있다면 제가 놓쳤을 수도 있으니 알려주세요📣
***
### 먼저 칭찬을 드리고 싶은 부분들이 있어요👏
- 세부적으로 SwiftLint를 적용해 혼자라도 코드 컨벤션을 통일시키려고 접목시킨점 정말 훌륭합니다💯
- 각 파일들을 용도에 맞게 그룹화 시키고 최대한 나누면서 구현하신점 최고💯
- 특히 의존성 주입을 많이 고민하시고 잘 적용해보셨더군요! 의존성 주입이 개념은 쉽지만 어려운? 부분인데 잘 접근하셨습니다💯
나중에 시간이 되신다면 의존성 주입을 잘 설계하고 사용하도록 도와준 써드파티 라이브러리가 있어요!
바로 [Swinject](https://github.com/Swinject/Swinject)인데 한번 보시면 도움이 될것 같아요🧐

### 궁금한점❓
- 코멘트로도 남겨두었지만 TestCase에서 통과하지 못하는 케이스가 하나 있더라구요!

### 산체스의 고민에 대한 저의 생각🤔
- 저는 지금도 프로토콜을 잘 사용하고 코드 구조 분리도 잘해주신것 같아요👍
다만 Request와 관련된 코드 및 폴더의 파일들이 상세히 나눠졌는데 파일명이나 메서드명만 보고 어떤 기능 혹은 정보를 담고 있는지 쉽게 파악될 수 있으면 더 좋을것 같습니다! 
세세하게 코드를 나누는것도 좋지만 가독성을 해친다면 공통된것은 합치는 방향도 있을것 같아요😀
- 테스트를 위해 기존과 다르게 타겟 분리해준 부분은 💯
더 쉽게 코드 분리가 되고 역할을 확실히 나눠주었네요! 멋집니다😃

### 이런 방향은 어떨까요🧐
- 지금도 코드가 역할별로 잘 나눠져 있지만 조금 더 욕심을 내보자면 에러 타입들도 에러별로 그룹핑을 시켜주는건 어떨까요?
후에 여러 에러 타입들이 생기고 한다면 분리하는것도 좋은 방향일거라 생각합니다!
지금 잘해주신 Manager 그룹핑 처럼요😁
- 나머지는 코멘트로 의견을 제시드립니다🙏🏻

### Github에 대한 Tip
- 아실수도 있지만 PR 오른쪽에 Reviewers에 저를 지정해주시고 새로 푸시하거나 수정후 재리뷰를 요청할때 
Re-request review 버튼을 통해 쉽게 요청하실 수 있습니다🥳
***
이번 Step1 많은 부분들을 적용하시느라 정말 수고 많으셨어요!
이번 스텝에서 필수 요구된 사항들은 확인이 되었고 산체스도 이번 플젝은 저번에 못다한 UI에 더 치중하고 싶으신것 같아
Request changes 대신 Comment만 남깁니다!
지금 이대로 다음 스텝으로 가도 괜찮지만 몇가지 코멘트들은 생각해보시고 적용하신다면 더 좋을것 같아요😀
수정을 하셔도 되고 "아니다 난 지금 코드가 더 좋다!" 하신다면 머지 요청 해주세요☺️

sanches37 committed:
refactor: r/n/r/n 을 하나의 상수로 사용하도록 수정
명시적으로 사용하도록 let doubleLineBreak = "\r\n\r\n" 로 사용

sanches37 committed:
refactor: 카멜케이스를 고려하여 상수명 변경
filename -> fimeName 으로 카멜케이스 고려하여 변경

sanches37 committed:
refactor: Data+extension 로 파일명 변경
기존 BaseTypeExtension 에서 Data+extension 로 파일명 변경

sanches37 committed:
refactor: commuteWithAPI 의 failure 타입 변경
commuteWithAPI 의 failure 타입을 Error -> NetworkError 로 변경

sanches37 committed:
test: 상수값 오류 수정하여 테스트 통과하도록 수정
let expectInputValue = "Items" -> let expectInputValue = "Item" 으로 변경

sanches37 committed:
refactor: 규정된 명칭 통일하여 네이밍 수정

sanches37 committed:
refactor: ItemsData 와 ItemData 네이밍 변경
ItemData 는 Product 로 변경
ItemsData 는 Products 로 하면 가독성이 떨어져서 ProductCollection 으로 변경

sanches37 committed:
refactor: 이미지 파일명 수정
이미지 파일명 Step1PostTestImage 으로 수정

sanches37 committed:
refactor: valuableMethod 네이밍 변경 및 상수값 변경
valuableHTTPMethod 으로 네이밍 변경
var valuableHTTPMethod: [APIMethod] 로 타입만 명시하는 걸로 변경

sanches37 committed:
refactor: debugPrint 및 주석처리 삭제

sanches37 committed:
refactor: assetFailed -> assetLoadFailed 로 네이밍 변경

sanches37 committed:
refactor: receivedDataAsset -> LoadedDataAsset 으로 변경

sanches37 committed:
refactor: APIMethod, ContentType enum 수정
APIMethod, ContentType CustomStringConvertible 채택 해제
APIMethod -> APIHTTPMethod 로 네이밍 변경

sanches37 committed:
refactor: fileName 난수처리 부분 Date()를 사용하게 수정

sanches37 committed:
refactor: swift 파일 폴더 재배치
enum, protocol, error 폴더를 만들어서 파일 재배치

sanches37 commented:
안녕하세요 그린 리뷰해주신 부분이 정답은 아니라고  적용할 부분만 바꾸라고 하셨는데, 가슴에 와닿는 내용들 뿐이었습니다 😀
제가 제대로 리뷰수정을 한건지 잘 모르겠네요.
업무하시느라 바쁘신 와중에 리뷰도 빠르게해주시고, 답변도 빨리해주셔서 정말 감사할 따름입니다.🙏🏼
더 수정할 부분이나 다음 스탭 진행도 열심히 임하겠습니다! 💪🏻

sanches37 committed:
refactor: dataTask error 처리 부분 if let 문으로 변경
guard error = nil -> if let _ = error 코드로 변경

sanches37 committed:
refactor: assetLoadFailed -> LoadAssetFailed 로 변경

sanches37 committed:
refactor: LoadAssetFailed 카멜케이스 적용하여 수정
카멜케이스 적용하여 loadAssetFailed 로 수정

sanches37 committed:
refactor: LoadedDataAsset 카멜케이스 적용하여 수정
LoadedDataAsset -> loadedDataAsset 로 수정

sanches37 committed:
refactor: valuabelHTTPMethod 네이밍 수정
valuableHTTPMethod -> applicableHTTPMethod 로 변경

GREENOVER commented on OpenMarket/OpenMarket/Controller/ViewController.swift:
```diff
@@ -7,12 +7,10 @@
 import UIKit
 
 class ViewController: UIViewController {
+    let networkManager = NetworkManager()
```
생각해보니 해당 프로퍼티가 다른곳에서 알 필요가 있을까요?🧐

sanches37 replied:
이걸 지우질 않았네요 😂 
step2 진행시 수정하겠습니다

GREENOVER replied:
잉 안지우셔도 됩니다! 접근제어자에 대해 의견 제시를 한 부분입니다🥲

GREENOVER reviewed (approved):
리뷰 반영을 적극적으로 해주셔서 감사합니다😀
저희의 conversation이 82개라니... 제가 너무 까탈스러웠던건 아닌가 싶네요😅
전보다 훠월씬 이해하기 편하고 구조적인 코드가 된것 같습니다!
Step1은 이정도면 충분한것 같아서 머지하겠습니다ㅎㅎ
하나만 더 코멘트를 달았는데 이전 리뷰들과 코멘트들은 이후 스텝을 진행하며 고민하고 적용해봐도 좋을것 같습니다!
고생많으셨습니다 산체스😁
이제 드디어 UI의 세계로 들어가게 되는군요..!
UI는 정말정말 쉬워보이지만 제일 어렵고 중요한 부분이니 충분히 고민하시면서 많이 학습하는 스텝이 되었으면 좋겠습니다:D
Step2도 화이팅입니다🥳
***
아 머지하려니 생각난게 있어서 말씀드립니다ㅎㅎ
머지에도 여러 옵션(squash / rebase 등)이 있는데 나중에 심심하면 어떤 차이가 있는지 알아보면 좋을것 같아요👍
현업에서 PR 및 커밋 그리고 branch 관리를 어떻게 하는지 엿볼수 있을거에요🤫
